lib/src/conflicts.rs

```diff
@@ -77,51 +79,75 @@ fn write_diff_hunks(hunks: &[DiffHunk], file: &mut dyn Write) -> std::io::Result
     Ok(())
 }
 
-async fn get_file_contents(store: &Store, path: &RepoPath, term: &Option<FileId>) -> ContentHunk {
+#[derive(Error, Debug)]
+pub enum MaterializationError {
+    #[error(transparent)]
+    Backend(#[from] BackendError),
+    #[error("Failed to read file contents for {} with id {}: {source}", path.as_internal_file_string(), id.hex())]
+    Read {
+        path: RepoPathBuf,
+        id: FileId,
+        source: std::io::Error,
+    },
+    #[error(transparent)]
+    Write(std::io::Error),
+}
+
+async fn get_file_contents(
+    store: &Store,
+    path: &RepoPath,
+    term: &Option<FileId>,
+) -> Result<ContentHunk, MaterializationError> {
     match term {
         Some(id) => {
             let mut content = vec![];
             store
                 .read_file_async(path, id)
-                .await
-                .unwrap()
+                .await?
                 .read_to_end(&mut content)
-                .unwrap();
-            ContentHunk(content)
+                .map_err(|err| MaterializationError::Read {
+                    path: path.to_owned(),
+                    id: id.clone(),
+                    source: err,
+                })?;
+            Ok(ContentHunk(content))
         }
         // If the conflict had removed the file on one side, we pretend that the file
         // was empty there.
-        None => ContentHunk(vec![]),
+        None => Ok(ContentHunk(vec![])),
     }
 }
 
 pub async fn extract_as_single_hunk(
     merge: &Merge<Option<FileId>>,
     store: &Store,
     path: &RepoPath,
-) -> Merge<ContentHunk> {
+) -> Result<Merge<ContentHunk>, MaterializationError> {
     let builder: MergeBuilder<ContentHunk> = futures::stream::iter(merge.iter())
         .then(|term| get_file_contents(store, path, term))
-        .collect()
-        .await;
-    builder.build()
+        .try_collect()
+        .await?;
+    Ok(builder.build())
 }
 
 pub async fn materialize(
     conflict: &MergedTreeValue,
     store: &Store,
     path: &RepoPath,
     output: &mut dyn Write,
-) -> std::io::Result<()> {
+) -> Result<(), MaterializationError> {
     if let Some(file_merge) = conflict.to_file_merge() {
         let file_merge = file_merge.simplify();
-        let content = extract_as_single_hunk(&file_merge, store, path).await;
-        materialize_merge_result(&content, output)
+        let content = extract_as_single_hunk(&file_merge, store, path).await?;
+        materialize_merge_result(&content, output).map_err(MaterializationError::Write)?;
```

Comment:
`materialize_tree_value_no_access_denied()` has to be updated to propagate "read" kind of errors.

Reply:
`materialize_tree_value_no_access_denied()` returns a `MaterializedTreeValue`, which has a `Box<dyn Reader>` in its `File` variant, so any failure to read the contents will only be known later. What am I missing?

Reply:
I meant this expectation is now invalid.

```rust
        Err(conflict) => {
            let mut contents = vec![];
            materialize(&conflict, store, path, &mut contents)
                .await
                .expect("Failed to materialize conflict to in-memory buffer");
```

Reply:
Good catch! I inlined `materialize()` in a newly added commit, which made it trivial to propagate the error.